src/Ganeti/HTools/Loader.hs

```diff
@@ -372,29 +373,49 @@ clearDynU cdata@(ClusterData _ _ il _ _) =
   let il2 = Container.map (\ inst -> inst {Instance.util = zeroUtil }) il
   in Ok cdata { cdInstances = il2 }
 
--- | Checks the cluster data for consistency.
-checkData :: Node.List -> Instance.List
-          -> ([String], Node.List)
-checkData nl il =
-    Container.mapAccum
-        (\ msgs node ->
-             let nname = Node.name node
-                 delta_mem = truncate (Node.tMem node)
-                             - Node.nMem node
-                             - Node.fMem node
-                             - nodeImem node il
-                 delta_dsk = truncate (Node.tDsk node)
-                             - Node.fDsk node
-                             - nodeIdsk node il
-                 newn = node `Node.setXmem` delta_mem
-                 umsg1 =
-                   if delta_mem > 512 || delta_dsk > 1024
-                      then printf "node %s is missing %d MB ram \
-                                  \and %d GB disk"
-                                  nname delta_mem (delta_dsk `div` 1024):msgs
-                      else msgs
-             in (umsg1, newn)
-        ) [] nl
+-- | Update cluster data to use static node memory on KVM.
+setStaticKvmNodeMem :: Node.List -- ^ Nodes to update
+                       -> Int -- ^ Static node size
+                       -> Node.List -- ^ Updated nodes
+setStaticKvmNodeMem nl static_node_mem =
+  let updateNM n = if Node.hypervisor n == Just Kvm
+                   then n { Node.nMem = static_node_mem }
+                   else n
+  in Container.map updateNM nl
+
+-- | Update node memory stat based on instance list.
+updateMemStat :: Node.Node -> Instance.List -> Node.Node
+updateMemStat node il =
+    let missing = truncate (Node.tMem node)
+                  - Node.nMem node
+                  - Node.fMem node
+                  - nodeImem node il
+    in node { Node.xMem = missing }
+
+-- | Check the cluster for memory/disk allocation consistency and update stats.
+updateMissing :: Node.List -- ^ All nodes in the cluster
+              -> Instance.List  -- ^ All instances in the cluster
+              -> Int -- ^ Static node memory for KVM
+              -> ([String], Node.List) -- ^ Pair of errors, update node list
+updateMissing nl il static_node_mem =
+    -- This overrides node mem on KVM as loaded from backend. Ganeti 2.17
+    -- handles this using obtainNodeMemory.
```

Comment:
Moved the condition. I don't think we have many choices where to call this function, bear in mind that it has to work for IAllocator as well that does not use loadExternalData in the generic case, but it does call updateMissing. Also we need to put it somewhere, where we have access to options. This sounded like the best possible compromise. Do you have any ideas?